plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java

```diff
@@ -560,6 +590,10 @@ void executeSingleUpload(final S3BlobStore blobStore, final String blobName, fin
             .storageClass(blobStore.getStorageClass())
             .acl(blobStore.getCannedACL())
             .overrideConfiguration(o -> o.addMetricPublisher(blobStore.getStatsMetricPublisher().putObjectMetricPublisher));
+
+        if (metadata != null && !metadata.isEmpty()) {
+            putObjectRequestBuilder = putObjectRequestBuilder.metadata(metadata);
```

Comment:
What happens if we assign null or empty metadata? Trying to understand if we really need the check in `if` condition.

Reply:
This if condition is not necessarily required here. But it's good to keep it as the function is also used for other files upload and assigning metadata every time even if it is null/empty can increase the request size.

Reply:
> But it's good to keep it as the function is also used for other files upload and assigning metadata every time even if it is null/empty increases the request size.

I am not sure about it. If we don't assign anything, what value of metadata is picked in the `PutObjectRequest`? Would the behaviour be same as assigning `null`?

Reply:
When we don't assign anything. It always create empty instance.
```
this.metadata = DefaultSdkAutoConstructMap.getInstance();
```


When we assign any value to metadata even if it is null/empty.  It tries to create a copy using

```
public final Builder metadata(Map<String, String> metadata) {
            this.metadata = MetadataCopier.copy(metadata);
            return this;
        }
```

This creates a copy of unmodifiable using method  - 

```
static Map<String, String> copy(Map<String, String> metadataParam) {
        Object map;
        if (metadataParam != null && !(metadataParam instanceof SdkAutoConstructMap)) {
            Map<String, String> modifiableMap = new LinkedHashMap();
            metadataParam.forEach((key, value) -> {
                modifiableMap.put(key, value);
            });
            map = Collections.unmodifiableMap(modifiableMap);
        } else {
            map = DefaultSdkAutoConstructMap.getInstance();
        }

        return (Map)map;
    }
```

Reply:
Thanks for fetching the details.